### PR TITLE
Text field length

### PR DESCRIPTION
**Pull request by stdavis**

### CODE REVIEW

File: src/forklift/core.py [hunk 1]
```diff
@@ -184,9 +184,13 @@ def abstract_type(type):
...
                 mismatching_fields.append('{}: source type of {} does not match destination type of {}'
                                           .format(source_fld.name, source_fld.type, destination_fld.type))
-            elif source_fld.type == 'String' and source_fld.length != destination_fld.length:
-                mismatching_fields.append('{}: source length of {} does not match destination length of {}'
-                                          .format(source_fld.name, source_fld.length, destination_fld.length))
+            elif source_fld.type == 'String':
+                if source_fld.length > 4000:
+                    log.warn('%s is longer than 4000 characters. Truncation may occur.', source_fld.name)
+                    source_fld.length = 4000
+                if source_fld.length != destination_fld.length:
```
Comment (steveoh) [inline src/forklift/core.py]: should this be an else if?

Comment (stdavis) [inline src/forklift/core.py]: No, because that would end up raising `ValidationException` and prevent the data from being updated.

Comment (steveoh) [inline src/forklift/core.py]: Given

```py
#: setup test
source_fld.length = 5000
source_fld.type = 'String'
destination_fld.length = 4001

#: back to code in question
if source_fld.length > 4000:
    log.warn('%s is longer than 4000 characters. Truncation may occur.', source_fld.name)
    source_fld.length = 4000
if source_fld.length != destination_fld.length:
   mismatching_fields.append('{}: source length of {} does not match destination length of {}'
                                               .format(source_fld.name, source_fld.length, destination_fld.length))
```
both if statements will be true and a `ValidationException` will be thrown.

Am I missing an assumption that destination_fld.length will never be greater than 4000?

Comment (stdavis) [inline src/forklift/core.py]: Good point. I was stuck on destination being no more than 4000...

File: src/forklift/core.py [hunk 2]
```diff
@@ -184,9 +185,19 @@ def abstract_type(type):
...
+            elif source_fld.type == 'String':
+                def truncate_field_length(field):
+                    if field.length > 4000:
+                        log.warn('%s is longer than 4000 characters. Truncation may occur.', field.name)
+                        return 4000
+                    else:
+                        return field.length
+
+                source_fld.length = truncate_field_length(source_fld)
+                destination_fld.length = truncate_field_length(destination_fld)
```
Comment (steveoh) [inline src/forklift/core.py]: if the destination length is longer, why should it get truncated?

Comment (stdavis) [inline src/forklift/core.py]: Just for standardization of the comparison in the line below? I don't know, really.

Comment (steveoh) [inline src/forklift/core.py]: can you think of a reason for checking source and destination lengths if you truncate the source?

**Comment by steveoh** [comment]
LGTM, might as well give it a try

**Comment by stdavis** [comment]
Merged.

Still working on getting rid of "*"...

**Pull request closed without merging**